### DISCUSSION TITLE
v1.1.317 has improved performance for packages in the numpy ecosystem

### DISCUSSION THREAD

**Discussion by has2k1** [microsoft/pyright#5496] [General]
Over the last couple of versions, I noticed that pyright was slowly and sometimes dramatically taking longer. A little profiling pointed to numpy as severely affecting performance and probably the main cause of the degraded results. 

Here is a snapshot of the trend. 

![pyright-performance](https://github.com/microsoft/pyright/assets/780341/23447401-2062-4ade-81cc-fa708a3a2caf)

[mizani](https://github.com/has2k1/mizani), [pandas](https://github.com/pandas-dev/pandas) and [plotnine](https://github.com/has2k1/plotnine) use a lot (relative to project size) of numpy. [py-shiny](https://github.com/rstudio/py-shiny) not so much.

**Comment by erictraut** [comment]
Thanks for sharing. I'm glad to see that my recent optimizations have had a measurable impact.

Out of curiosity, were all of these stats collected using the same version of numpy? I've noticed that recent versions of numpy have made increased usage of recursive protocols. My recent optimizations primarily targeted those.

**Reply by has2k1** [reply]
No they weren't using the same version.

The versions are

```
plotnine: numpy-1.24.3
mizani: numpy-1.24.2
py-shiny: numpy-1.25.1
pandas: numpy-1.25.1
```
